ttnn/cpp/ttnn/tensor/types.hpp

```diff
@@ -424,35 +424,38 @@ struct MultiDeviceHostStorage {
         std::vector<OwnedBuffer> buffers;
         std::vector<Shape> shapes;
         mutable std::mutex mtx;
+
+        friend void swap(MultiDeviceHostStorage& first, MultiDeviceHostStorage& second) {
+            std::scoped_lock lock(first.mtx, second.mtx);
+            // enable ADL (not necessary, but good practice)
+            using std::swap;
+
+            swap(first.strategy, second.strategy);
+            swap(first.buffers, second.buffers);
+            swap(first.shapes, second.shapes);
+        }
+
         MultiDeviceHostStorage() = default;
         MultiDeviceHostStorage(DistributedTensorConfig strategy_, std::vector<OwnedBuffer> buffers_, std::vector<Shape> shapes_) : strategy(strategy_), buffers(buffers_), shapes(shapes_) {}
         MultiDeviceHostStorage(MultiDeviceHostStorage &&other) {
-            std::lock_guard<std::mutex> lock(mtx);
-            strategy = other.strategy;
-            buffers = other.buffers;
-            shapes = other.shapes;
+            swap(*this, other);
         }
-
+        // unfotunately we need to have this code written manually.
         MultiDeviceHostStorage(const MultiDeviceHostStorage &other) {
-            std::lock_guard<std::mutex> lock(mtx);
+            std::scoped_lock lock(mtx, other.mtx);
             strategy = other.strategy;
             buffers = other.buffers;
             shapes = other.shapes;
         }
 
         MultiDeviceHostStorage &operator=(const MultiDeviceHostStorage &other) {
-            std::lock_guard<std::mutex> lock(mtx);
-            strategy = other.strategy;
-            buffers = other.buffers;
-            shapes = other.shapes;
+            MultiDeviceHostStorage temp(other);
+            swap(*this, temp);
```

Comment:
This could even be
```suggestion
            *this = MultiDeviceHostStorage(other);
```
which delegates to the move assignment operator.

Reply:
Hmm cannot decide which one is better. 
1 line vs 2 lines
But swap call looks easier to understand.

Reply:
This one isn't a request, it's just a suggestion, so feel free to dismiss if the existing code makes more sense to you.

Reply:
I always used my approach and never saw yours. Never saw it before. Interesting idea.
I think explicit swap call looks better for me.